internal/serialization/schema_service.go

```diff
@@ -19,21 +16,67 @@
 
 package serialization
 
+import (
+	"context"
+	"sync"
+
+	pubserialization "github.com/hazelcast/hazelcast-go-client/serialization"
+)
+
+type SchemaMsg struct {
+	ID         int64
+	ResponseCh chan *Schema
+}
+
 type SchemaService struct {
-	schemaMap map[int64]Schema
+	schemaMap map[int64]*Schema
+	mu        *sync.RWMutex
+	ch        chan<- SchemaMsg
 }
 
-func NewSchemaService() *SchemaService {
+func NewSchemaService(cfg pubserialization.CompactConfig, ch chan<- SchemaMsg) *SchemaService {
 	return &SchemaService{
-		schemaMap: make(map[int64]Schema),
+		schemaMap: MakeSchemasFromConfig(cfg),
+		mu:        &sync.RWMutex{},
+		ch:        ch,
 	}
 }
 
-func (s *SchemaService) Get(schemaId int64) (Schema, bool) {
-	schema, ok := s.schemaMap[schemaId]
+func (s *SchemaService) Schemas() []*Schema {
+	s.mu.RLock()
+	schemas := make([]*Schema, 0, len(s.schemaMap))
+	for _, v := range s.schemaMap {
+		schemas = append(schemas, v)
+	}
+	s.mu.RUnlock()
+	return schemas
+}
+
+func (s *SchemaService) Get(ctx context.Context, schemaId int64) (schema *Schema, ok bool) {
+	// TODO: return error
+	s.mu.RLock()
+	schema, ok = s.schemaMap[schemaId]
+	s.mu.RUnlock()
+	if !ok {
+		rch := make(chan *Schema)
+		s.ch <- SchemaMsg{ID: schemaId, ResponseCh: rch}
+		select {
+		case schema = <-rch:
+			ok = schema != nil
+			s.putLocal(schema)
```

Comment:
if schema is null, wouldn't this call throw?

Reply:
It would, fixed at: https://github.com/hazelcast/hazelcast-go-client/pull/910/commits/05d7f6629a2d0b144925608c3f3e9e07b5f9a570